src/lib/Lower.hs

```diff
@@ -0,0 +1,130 @@
+-- Copyright 2022 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+module Lower (lowerFullySequential) where
+
+import Data.Map.Strict qualified as M
+import Control.Monad
+import Unsafe.Coerce
+
+import Types.Core
+import Types.Primitives
+
+import Name
+import MTL1
+import Core
+import Builder
+import QueryType
+import GenericTraversal
+
+lowerFullySequential :: EnvReader m => Block n -> m n (Block n)
+lowerFullySequential b = liftM fst $ liftGenericTraverserM LFS $ traverseGenericE b
+
+data LFS (n::S) = LFS
+type LowerM = GenericTraverserM LFS
+instance SinkableE LFS where
+  sinkingProofE _ LFS = LFS
+instance HoistableState LFS where
+  hoistState _ _ LFS = LFS
+  {-# INLINE hoistState #-}
+
+instance GenericTraverser LFS where
+  traverseExpr expr = case expr of
+    Hof (For dir ixTy (Lam body)) -> traverseFor Nothing dir ixTy body
+    _ -> traverseExprDefault expr
+
+type Dest = Atom
+
+traverseFor :: Emits o => Maybe (Dest o) -> ForAnn -> Atom i -> LamExpr i -> LowerM i o (Expr o)
+traverseFor maybeDest dir ixTy lam@(LamExpr (LamBinder ib ty arr eff) body) = do
+  initDest <- case maybeDest of
+    Just  d -> return d
+    Nothing -> emitOp . AllocDest =<< getTypeSubst (Hof $ For dir ixTy (Lam lam))
+  destTy <- getType initDest
+  ty' <- substM ty
+  ixTy' <- substM ixTy
+  eff' <- substM $ ignoreHoistFailure $ hoist ib eff
+  body' <- buildLam noHint arr (PairTy ty' destTy) eff' \b' -> do
+    (i, dest) <- fromPair $ Var b'
+    idest <- emitOp $ IndexRef dest i
+    extendSubst (ib @> SubstVal i) $ traverseBlockWithDest idest body
+    return UnitVal
+  let seqHof = Hof $ Seq dir ixTy' initDest body'
+  case maybeDest of
+    Just _  -> return seqHof
+    Nothing -> Op . Freeze . Var <$> emit seqHof
+
+-- Destination-passing traversals
+
+type AtomNameMap (i'::S) (o::S) = M.Map (AtomName i') (Dest o)
+type DestAssignment (i'::S) (o::S) = AtomNameMap i' o
+
+lookupDest :: DestAssignment i' o -> AtomName i' -> Maybe (Dest o)
+lookupDest = flip M.lookup
+
+decomposeDest :: Emits o => Dest o -> Atom i' -> LowerM i o (Maybe (DestAssignment i' o))
+decomposeDest dest = \case
+  Var v -> return $ Just $ M.singleton v dest
```

Comment:
I think a TODO comment here about other intended decompositions would help make the rest of the code clearer.

Reply:
Good point, I'll add it tomorrow!